packages/service-library/src/servicelib/aiohttp/monitoring.py

```diff
@@ -38,105 +42,194 @@
 #
 
 
-# TODO: the endpoint label on the http_requests_total Counter is a candidate to be removed. as endpoints also contain all kind of UUIDs
+kSTART_TIME = f"{__name__}.start_time"
+kREQUEST_COUNT = f"{__name__}.request_count"
+kINFLIGHTREQUESTS = f"{__name__}.in_flight_requests"
+kRESPONSELATENCY = f"{__name__}.in_response_latency"
+kCANCEL_COUNT = f"{__name__}.cancel_count"
+
+kCOLLECTOR_REGISTRY = f"{__name__}.collector_registry"
+kPROCESS_COLLECTOR = f"{__name__}.collector_process"
+kPLATFORM_COLLECTOR = f"{__name__}.collector_platform"
+kGC_COLLECTOR = f"{__name__}.collector_gc"
+
 
+def get_collector_registry(app: web.Application) -> CollectorRegistry:
+    return app[kCOLLECTOR_REGISTRY]
+
+
+async def metrics_handler(request: web.Request):
+    registry = get_collector_registry(request.app)
+
+    # NOTE: Cannot use ProcessPoolExecutor because registry is not pickable
+    result = await request.loop.run_in_executor(
+        None, prometheus_client.generate_latest, registry
+    )
+    response = web.Response(body=result)
+    response.content_type = CONTENT_TYPE_LATEST
+    return response
 
-async def metrics_handler(_request: web.Request):
-    # TODO: prometheus_client.generate_latest blocking! no asyhc solutin?
-    # TODO: prometheus_client access to a singleton registry! can be instead created and pass to every metric wrapper
-    resp = web.Response(body=prometheus_client.generate_latest())
-    resp.content_type = CONTENT_TYPE_LATEST
-    return resp
 
+EnterMiddlewareCB = Callable[[web.Request], Awaitable[None]]
+ExitMiddlewareCB = Callable[[web.Request, web.StreamResponse], Awaitable[None]]
 
-def middleware_factory(app_name):
+
+def middleware_factory(
+    app_name: str,
+    enter_middleware_cb: Optional[EnterMiddlewareCB],
```

Comment:
it is applied to storage. Or do you mean the director-v0?

for the second part, I actually ended up doing it this way because of the diagnostics part in the webserver. I would be in favor of separating the diagnostics part (the module that sense if the webserver is behaving correctly) from the monitoring part.
Anyway your points make sense, I will check these cases.

Reply:
ah, of course ... it is already there. I just did not see any changed file of storage. NO, director-v0 stays as it is ... do not open pandoras box please :-D

yes, diagnostics should be separated from monitoring. The former is a "conclusion from the data produced in the former"

Reply:
@pcrespov : I will ask for a new review from you. I added a new function to catch and log exceptions in logging_utils.py